content/blog/2023-03-23-when-a-date-breaks-booting-the-kernel.md

```diff
@@ -0,0 +1,213 @@
+```
+
+There are some new entries here: notably `init/version*` and `usr/initramfs*`. Binary searching these files results in a single culprit: `usr/initramfs_data.cpio`. This is quite fitting, as the `.cpio` file is an archive defining a filesystem layout, [much like `.tar` files](https://docs.kernel.org/filesystems/ramfs-rootfs-initramfs.html?highlight=initramfs#why-cpio-rather-than-tar). This file is actually embedded into the kernel image, and loaded as a bare-bones shim filesystem when the user doesn't provide their own initramfs[^1].
+
+[^1]: initramfs, or initrd for the older format, are specific kinds of CPIO archives. They are intended to be loaded as the initial filesystem of a booted kernel, typically in preparation
+
+So it would make sense that if the CPIO archive wasn't being rebuilt, then the initial filesystem wouldn't change. And it would make sense for the initial filesystem to be causing mount issues of the proper root disk filesystem.
+
+This just leaves the question of how `KBUILD_BUILD_TIMESTAMP` is breaking the CPIO archive. And it's around this time that a third developer Andrew I'd roped into this bug hunt pointed out that the generator script for this CPIO archive was passing the `KBUILD_BUILD_TIMESTAMP` to `date`. Whoop, we've found the murder weapon[^2]!
+
+[^2]: Hindsight again would suggest it was obvious to look here because it shows up when searching for `KBUILD_BUILD_TIMESTAMP`. I unfortunately wasn't familiar with the `usr/` source folder initially, and focused on the core kernel components too much earlier. Oh well, we found it eventually.
+
+The persistence bug could be explained now: because the script was only using `KBUILD_BUILD_TIMESTAMP` internally, `make` had no way of knowing that the archive generation depended on this variable. So even when I changed the variable to a valid value, `make` didn't know to rebuild the corrupt archive. Let's now get back to the main issue: why boot panics.
+
+
+## Solving the case
+
+Following along the CPIO generation script, the `KBUILD_BUILD_TIMESTAMP` variable is turned into a timestamp by `date -d"$KBUILD_BUILD_TIMESTAMP" +%s`. Testing this in the shell with `0000-01-01` we get this (somewhat amusing, but also painful) result
+```text
+date -d"$KBUILD_BUILD_TIMESTAMP" +%s
+-62167255492
+```
+
+This timestamp is then passed to a C program that assigns it to a variable `default_mtime`. Looking over the source, it seems this variable is used to set the `mtime` field on the files in the CPIO archive. The timestamp is stored as an `int64_t`. That's 64 bits of data, up to 16 hexadecimal characters. And yes, that's relevant: CPIO stores the `mtime` (and all other numerical fields) as ASCII hexadecimal characters. The `sprintf()` call that ultimately embeds the timestamp uses the `%08lX` format specifier. This formats a `long` as hexadecimal, padded to at least 8 characters. Hang on... ***at least*** 8 characters? What if our timestamp happens to be more?
```

Comment:
iirc the timestamp is read in as a time_t that is typedefed to be the same as an int64_t

Reply:
you should explicitly say that the CPIO date format is a 32-bit *unsigned* integer (seconds since epoch) represented as hex